File: internal/ollama/service.go
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package ollama

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json/v2"
	"errors"
	"fmt"
	"net/netip"
	"slices"
	"strings"
	"time"

	"github.com/containerd/errdefs"
	"github.com/moby/moby/api/types/container"
	"github.com/moby/moby/api/types/mount"
	"github.com/moby/moby/api/types/network"
	"github.com/moby/moby/api/types/volume"
	"github.com/moby/moby/client"
)

const ownerLabel = "nemoclaw.nvidia.com/uid"
const operationLabel = "nemoclaw.nvidia.com/generation"

var errPartialCreate = errors.New("Ollama container is absent but persistent storage remains; reconcile the recorded create or inspect the deployment")

const specLabel = "nemoclaw.nvidia.com/ollama-spec"

type ServiceSpec struct {
	Name, Owner, Generation, Image, Network, BindAddress string
}

type Service struct {
	Spec    ServiceSpec
	ID      string
	Running bool
}

type Docker struct{ API *client.Client }

func NewDocker(endpoint string) (*Docker, error) {
	// The first experiment deliberately selects a local engine. Remote engine
	// authentication and Windows/Podman topology require their own qualification.
	if !strings.HasPrefix(endpoint, "unix:///") {
		return nil, errors.New("managed Ollama requires an explicit local Unix engine socket")
	}
	c, err := client.New(client.WithHost(endpoint))
	if err != nil {
		return nil, errors.New("cannot configure Docker engine client")
	}
	return &Docker{API: c}, nil
}

func (d *Docker) Close() error { return d.API.Close() }

// Bound validates the physical parent before accessing its unauthenticated API.
func (d *Docker) Bound(ctx context.Context, id, endpoint string) (*Service, error) {
	parts := strings.Split(id, "/")
	if len(parts) != 3 {
		return nil, errors.New("invalid Ollama physical binding")
	}
	c, err := d.API.ContainerInspect(ctx, parts[1], client.ContainerInspectOptions{})
	if err != nil || c.Container.Config == nil || c.Container.HostConfig == nil {
		return nil, errors.New("bound Ollama container observation failed")
	}
	spec := ServiceSpec{Name: strings.TrimPrefix(c.Container.Name, "/"), Owner: c.Container.Config.Labels[ownerLabel], Generation: c.Container.Config.Labels[operationLabel], Image: c.Container.Config.Image, Network: string(c.Container.HostConfig.NetworkMode), BindAddress: strings.TrimSuffix(strings.TrimPrefix(endpoint, "http://"), "/v1")}
	s, err := d.Observe(ctx, spec)
	if err != nil {
		return nil, err
	}
	if s == nil || s.ID != id {
		return nil, errors.New("Ollama target or storage identity changed")
	}
	return s, nil
}
func (s ServiceSpec) Volume() string { return s.Name + "-models" }
func (s ServiceSpec) labels() map[string]string {
	b, _ := json.Marshal(s)
	h := sha256.Sum256(b)
	return map[string]string{ownerLabel: s.Owner, operationLabel: s.Generation, specLabel: hex.EncodeToString(h[:])}
}

func (d *Docker) Preflight(ctx context.Context, want ServiceSpec, id string) error {
	s, err := d.Observe(ctx, want)
	if id != "" {
		if err != nil || s == nil || s.ID != id {
			return errors.New("bound Ollama identity or storage is unavailable; automatic replacement is forbidden")
		}
		return nil
	}
	if !errors.Is(err, errPartialCreate) {
		return err
	}
	v, err := d.API.VolumeInspect(ctx, want.Volume(), client.VolumeInspectOptions{})
	if err != nil || v.Volume.CreatedAt == "" || v.Volume.Driver != "local" || len(v.Volume.Options) != 0 {
		return errors.New("incomplete retained Ollama storage")
	}
	for k, value := range want.labels() {
		if value == "" || v.Volume.Labels[k] != value {
			return errors.New("Ollama storage does not match the retained create operation")
		}
	}
	return nil
}

func (d *Docker) Observe(ctx context.Context, want ServiceSpec) (*Service, error) {
	ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
	defer cancel()
	info, err := d.API.Info(ctx, client.InfoOptions{})
	if err != nil || info.Info.ID == "" {
		return nil, errors.New("Docker target identity is unavailable")
	}
	c, ce := d.API.ContainerInspect(ctx, want.Name, client.ContainerInspectOptions{})
	v, ve := d.API.VolumeInspect(ctx, want.Volume(), client.VolumeInspectOptions{})
	if errdefs.IsNotFound(ce) && errdefs.IsNotFound(ve) {
		return nil, nil
	}
	if errdefs.IsNotFound(ce) && ve == nil {
		return nil, errPartialCreate
	}
	if ce != nil || ve != nil {
		return nil, errors.New("Ollama container or storage observation failed; absence is unconfirmed")
	}
	if err = verify(want, c.Container, v.Volume); err != nil {
		return nil, err
	}
	return &Service{Spec: want, ID: info.Info.ID + "/" + c.Container.ID + "/" + v.Volume.CreatedAt, Running: c.Container.State.Running}, nil
}

func verify(want ServiceSpec, c container.InspectResponse, v volume.Volume) error {
	if c.ID == "" || c.Config == nil || c.HostConfig == nil || c.State == nil || v.CreatedAt == "" {
		return errors.New("incomplete Docker resource metadata")
	}
	for k, value := range want.labels() {
		if value == "" || c.Config.Labels[k] != value || v.Labels[k] != value {
			return errors.New("Ollama ownership, operation identity, or configuration conflict")
		}
	}
	bind, err := netip.ParseAddrPort(want.BindAddress)
	if err != nil {
		return errors.New("invalid Ollama binding")
	}
	port, _ := network.ParsePort("11434/tcp")
	pb := c.HostConfig.PortBindings[port]
	if strings.TrimPrefix(c.Name, "/") != want.Name || c.Config.Image != want.Image ||
		string(c.HostConfig.NetworkMode) != want.Network || c.HostConfig.Privileged ||
		!slices.Contains(c.HostConfig.CapDrop, "ALL") || !slices.Contains(c.HostConfig.SecurityOpt, "no-new-privileges") ||
		!slices.Equal(c.Config.Entrypoint, []string{"/bin/ollama"}) || !slices.Equal(c.Config.Cmd, []string{"serve"}) ||
		len(c.HostConfig.PortBindings) != 1 || len(pb) != 1 || pb[0].HostIP != bind.Addr() || pb[0].HostPort != fmt.Sprint(bind.Port()) ||
		v.Name != want.Volume() || v.Driver != "local" || len(v.Options) != 0 {
		return errors.New("Ollama container configuration drift requires inspection")
	}
	if len(c.Mounts) != 1 || c.Mounts[0].Type != mount.TypeVolume || c.Mounts[0].Name != want.Volume() || c.Mounts[0].Destination != "/root/.ollama" || !c.Mounts[0].RW {
		return errors.New("Ollama persistent storage binding drifted")
	}
	return nil
}

// Ensure reconciles only the exact retained operation token. Configuration and
// runtime effects remain distinct from readiness of the HTTP/model endpoint.
func (d *Docker) Ensure(ctx context.Context, want ServiceSpec, id string) (*Service, error) {
	got, err := d.Observe(ctx, want)
	if err != nil && !errors.Is(err, errPartialCreate) {
		return nil, err
	}
	if got != nil {
		if id != "" && id != got.ID {
			return nil, errors.New("Ollama physical identity changed")
		}
		if !got.Running {
			// Recheck by immutable container ID just before starting.
			if _, err = d.API.ContainerStart(ctx, strings.Split(got.ID, "/")[1], client.ContainerStartOptions{}); err != nil {
				return nil, errors.New("Ollama start outcome is unknown; reapply to reconcile")
			}
		}
		return d.Observe(ctx, want)
	}
	if id != "" {
		return nil, errors.New("bound Ollama resource is unavailable; automatic recreation is forbidden")
	}
	// An interrupted initial create can leave only the named volume. Verify its
	// retained random operation token before continuing that same create.
	v, ve := d.API.VolumeInspect(ctx, want.Volume(), client.VolumeInspectOptions{})
	if ve == nil {
		for k, value := range want.labels() {
			if value == "" || v.Volume.Labels[k] != value {
				return nil, errors.New("existing Ollama storage belongs to another operation")
			}
		}
		if v.Volume.Driver != "local" || len(v.Volume.Options) != 0 {
			return nil, errors.New("unexpected Ollama volume configuration")
		}
		if _, ce := d.API.ContainerInspect(ctx, want.Name, client.ContainerInspectOptions{}); !errdefs.IsNotFound(ce) {
			return nil, errors.New("Ollama create cannot reconcile container observation")
		}
	} else if !errdefs.IsNotFound(ve) || err != nil {
		return nil, errors.New("Ollama create requires complete engine observations")
	}
	if _, err = d.API.ImageInspect(ctx, want.Image); errdefs.IsNotFound(err) {
		pull, e := d.API.ImagePull(ctx, want.Image, client.ImagePullOptions{})
		if e != nil {
			return nil, errors.New("Ollama image pull failed")
		}
		e = pull.Wait(ctx)
		pull.Close()
		if e != nil {
			return nil, errors.New("Ollama image pull incomplete")
		}
	} else if err != nil {
		return nil, errors.New("Ollama image observation failed")
	}
	if errdefs.IsNotFound(ve) {
		if _, err = d.API.VolumeCreate(ctx, client.VolumeCreateOptions{Name: want.Volume(), Driver: "local", Labels: want.labels()}); err != nil {
			return nil, errors.New("Ollama volume create outcome unknown; retain operation and reapply")
		}
	}
	// Inspect after create: VolumeCreate can return an existing named volume.
	v, err = d.API.VolumeInspect(ctx, want.Volume(), client.VolumeInspectOptions{})
	if err != nil {
		return nil, errors.New("cannot verify newly created Ollama storage")
	}
	for k, value := range want.labels() {
		if v.Volume.Labels[k] != value {
			return nil, errors.New("Ollama storage collision")
		}
	}
	bind, err := netip.ParseAddrPort(want.BindAddress)
	if err != nil {
		return nil, err
	}
	port, _ := network.ParsePort("11434/tcp")
	c, err := d.API.ContainerCreate(ctx, client.ContainerCreateOptions{Name: want.Name,
		Config: &container.Config{Image: want.Image, Labels: want.labels(), Entrypoint: []string{"/bin/ollama"}, Cmd: []string{"serve"}},
		HostConfig: &container.HostConfig{NetworkMode: container.NetworkMode(want.Network),
			Mounts:       []mount.Mount{{Type: mount.TypeVolume, Source: want.Volume(), Target: "/root/.ollama"}},
			PortBindings: network.PortMap{port: {{HostIP: bind.Addr(), HostPort: fmt.Sprint(bind.Port())}}},
			CapDrop:      []string{"ALL"}, SecurityOpt: []string{"no-new-privileges"}},
	})
	if err != nil || c.ID == "" {
		return nil, errors.New("Ollama container create outcome unknown; retain operation and reapply")
	}
	if _, err = d.API.ContainerStart(ctx, c.ID, client.ContainerStartOptions{}); err != nil {
		return nil, errors.New("Ollama start outcome unknown; retain operation and reapply")
	}
	return d.Observe(ctx, want)
}

File: crates/nemoclaw-sdk/src/ollama/service.rs
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
use crate::{
    Error, ObservationError,
    docker::{Engine, remote},
    managed::{GENERATION_LABEL, OWNER_LABEL},
};
use bollard::{
    models::{ContainerCreateBody, ContainerInspectResponse, Volume, VolumeCreateRequest},
    query_parameters::{CreateContainerOptions, RemoveContainerOptions},
};
use serde::{Deserialize, Serialize};
use serde_json::json;
use sha2::{Digest, Sha256};
use std::{collections::HashMap, net::SocketAddr, time::Duration};
const SPEC_LABEL: &str = "nemoclaw.nvidia.com/ollama-spec";
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase", deny_unknown_fields)]
pub struct ServiceSpec {
    pub name: String,
    pub owner: String,
    pub generation: String,
    pub image: String,
    pub network: String,
    pub bind_address: String,
}
#[derive(Clone, Debug)]
pub struct Service {
    pub spec: ServiceSpec,
    pub id: String,
    pub running: bool,
}
impl ServiceSpec {
    pub fn volume(&self) -> String {
        format!("{}-models", self.name)
    }
    pub fn validate(&self) -> Result<(), Error> {
        let bind = self
            .bind_address
            .parse::<SocketAddr>()
            .map_err(|_| Error::Conflict("invalid Ollama binding"))?;
        if !regex::Regex::new(r"^nc-[a-f0-9]{16}-ollama$")
            .unwrap()
            .is_match(&self.name)
            || self.owner.is_empty()
            || self.generation.is_empty()
            || !regex::Regex::new(r"^ollama/ollama@sha256:[a-f0-9]{64}$")
                .unwrap()
                .is_match(&self.image)
            || !regex::Regex::new(r"^[a-z0-9][a-z0-9_-]*$")
                .unwrap()
                .is_match(&self.network)
            || bind.port() == 0
            || !(bind.ip().is_loopback()
                || match bind.ip() {
                    std::net::IpAddr::V4(ip) => ip.is_private(),
                    std::net::IpAddr::V6(ip) => ip.is_unique_local(),
                })
        {
            return Err(Error::Conflict(
                "Ollama requires owned, pinned configuration and a private bind address",
            ));
        }
        Ok(())
    }
    fn labels(&self) -> Result<HashMap<String, String>, Error> {
        let bytes = serde_json::to_vec(self)
            .map_err(|_| Error::State("cannot encode Ollama specification"))?;
        let digest = Sha256::digest(bytes)
            .iter()
            .map(|b| format!("{b:02x}"))
            .collect();
        Ok([
            (OWNER_LABEL.into(), self.owner.clone()),
            (GENERATION_LABEL.into(), self.generation.clone()),
            (SPEC_LABEL.into(), digest),
        ]
        .into())
    }
    fn container(&self) -> Result<ContainerCreateBody, Error> {
        let bind = self
            .bind_address
            .parse::<SocketAddr>()
            .map_err(|_| Error::Conflict("invalid Ollama binding"))?;
        serde_json::from_value(json!({"Image":self.image,"Labels":self.labels()?,"Entrypoint":["/bin/ollama"],"Cmd":["serve"],"HostConfig":{"NetworkMode":self.network,"Mounts":[{"Type":"volume","Source":self.volume(),"Target":"/root/.ollama"}],"PortBindings":{"11434/tcp":[{"HostIp":bind.ip().to_string(),"HostPort":bind.port().to_string()}]},"CapDrop":["ALL"],"SecurityOpt":["no-new-privileges"]}})).map_err(|_|Error::State("invalid compiled Ollama container"))
    }
    fn verify_volume(&self, volume: &Volume) -> Result<(), Error> {
        if volume.name != self.volume()
            || volume.driver != "local"
            || !volume.options.is_empty()
            || volume.created_at.as_ref().is_none_or(String::is_empty)
        {
            return Err(Error::Conflict(
                "Ollama persistent volume configuration is incomplete or changed",
            ));
        }
        for (key, value) in self.labels()? {
            if volume.labels.get(&key) != Some(&value) {
                return Err(ObservationError::BindingMismatch.into());
            }
        }
        Ok(())
    }
    fn verify(&self, container: &ContainerInspectResponse, volume: &Volume) -> Result<bool, Error> {
        self.verify_volume(volume)?;
        let config = container
            .config
            .as_ref()
            .ok_or(ObservationError::Incomplete)?;
        let host = container
            .host_config
            .as_ref()
            .ok_or(ObservationError::Incomplete)?;
        let running = container
            .state
            .as_ref()
            .and_then(|s| s.running)
            .ok_or(ObservationError::Incomplete)?;
        let labels = config.labels.as_ref().ok_or(ObservationError::Incomplete)?;
        for (key, value) in self.labels()? {
            if labels.get(&key) != Some(&value) {
                return Err(ObservationError::BindingMismatch.into());
            }
        }
        let launch = self.container()?;
        let expected = launch.host_config.unwrap();
        if container.id.as_ref().is_none_or(String::is_empty)
            || container.name.as_deref().map(|n| n.trim_start_matches('/')) != Some(&self.name)
            || config.image.as_deref() != Some(&self.image)
            || config.entrypoint != launch.entrypoint
            || config.cmd != launch.cmd
            || host.network_mode != expected.network_mode
            || host.privileged.unwrap_or(false)
            || !host
                .cap_drop
                .as_ref()
                .is_some_and(|values| values.iter().any(|v| v == "ALL"))
            || !host
                .security_opt
                .as_ref()
                .is_some_and(|values| values.iter().any(|v| v == "no-new-privileges"))
            || host.port_bindings != expected.port_bindings
        {
            return Err(Error::Conflict(
                "Ollama container configuration drift requires inspection",
            ));
        }
        let mounts = container
            .mounts
            .as_ref()
            .ok_or(ObservationError::Incomplete)?;
        if mounts.len() != 1
            || mounts[0].name.as_deref() != Some(&self.volume())
            || mounts[0].destination.as_deref() != Some("/root/.ollama")
            || mounts[0].rw != Some(true)
            || mounts[0].typ.as_deref() != Some("volume")
        {
            return Err(Error::Conflict("Ollama persistent storage binding drifted"));
        }
        Ok(running)
    }
}
impl Engine {
    /// Observe storage independently of whether its service is running or present.
    pub async fn observe_ollama_storage(
        &self,
        spec: &ServiceSpec,
        id: &str,
    ) -> Result<Option<String>, Error> {
        spec.validate()?;
        let info = self.info().await?;
        let Some(volume) = self.volume(&spec.volume()).await? else {
            if !id.is_empty() {
                return Err(Error::Conflict(
                    "bound Ollama storage is absent; recreation forbidden",
                ));
            }
            return Ok(None);
        };
        spec.verify_volume(&volume)?;
        let physical = format!(
            "{}/{}/{}",
            info.id.unwrap(),
            spec.volume(),
            volume.created_at.unwrap()
        );
        if !id.is_empty() && id != physical {
            return Err(ObservationError::BindingMismatch.into());
        }
        Ok(Some(physical))
    }
    pub async fn ensure_ollama_storage(
        &self,
        spec: &ServiceSpec,
        id: &str,
    ) -> Result<String, Error> {
        if let Some(id) = self.observe_ollama_storage(spec, id).await? {
            return Ok(id);
        }
        self.api
            .create_volume(VolumeCreateRequest {
                name: Some(spec.volume()),
                driver: Some("local".into()),
                labels: Some(spec.labels()?),
                ..Default::default()
            })
            .await
            .map_err(|error| remote(&error))?;
        self.observe_ollama_storage(spec, id)
            .await?
            .ok_or(ObservationError::Incomplete.into())
    }
    pub async fn observe_ollama_removal(
        &self,
        spec: &ServiceSpec,
        id: &str,
    ) -> Result<Option<Service>, Error> {
        let parts: Vec<_> = id.split('/').collect();
        if parts.len() != 3 || parts.iter().any(|p| p.is_empty()) {
            return Err(ObservationError::Incomplete.into());
        }
        self.observe_ollama_storage(
            spec,
            &format!("{}/{}/{}", parts[0], spec.volume(), parts[2]),
        )
        .await?;
        if self.container(&spec.name).await?.is_none() {
            return Ok(None);
        }
        self.observe_ollama(spec, id).await
    }
    pub async fn remove_ollama(&self, spec: &ServiceSpec, id: &str) -> Result<(), Error> {
        if let Some(service) = self.observe_ollama_removal(spec, id).await? {
            let container = service
                .id
                .split('/')
                .nth(1)
                .ok_or(ObservationError::Incomplete)?;
            self.api
                .remove_container(
                    container,
                    Some(RemoveContainerOptions {
                        force: true,
                        v: false,
                        ..Default::default()
                    }),
                )
                .await
                .map_err(|error| remote(&error))?;
        }
        Ok(())
    }
    pub async fn observe_ollama(
        &self,
        spec: &ServiceSpec,
        id: &str,
    ) -> Result<Option<Service>, Error> {
        spec.validate()?;
        let work = async {
            let info = self.info().await?;
            let container = self.container(&spec.name).await?;
            let volume = self.volume(&spec.volume()).await?;
            let Some(container) = container else {
                if let Some(volume) = volume {
                    spec.verify_volume(&volume)?;
                    return Err(Error::PartialRuntime);
                }
                if !id.is_empty() {
                    return Err(Error::Conflict(
                        "bound Ollama is absent; automatic recreation is forbidden",
                    ));
                }
                return Ok(None);
            };
            let volume = volume.ok_or(Error::Conflict(
                "Ollama storage observation failed; absence is unconfirmed",
            ))?;
            let running = spec.verify(&container, &volume)?;
            let physical = format!(
                "{}/{}/{}",
                info.id.unwrap(),
                container.id.unwrap(),
                volume.created_at.unwrap()
            );
            if !id.is_empty() && id != physical {
                return Err(ObservationError::BindingMismatch.into());
            }
            Ok(Some(Service {
                spec: spec.clone(),
                id: physical,
                running,
            }))
        };
        tokio::time::timeout(Duration::from_secs(20), work)
            .await
            .map_err(|_| ObservationError::Transport)?
    }
    pub async fn preflight_ollama(&self, spec: &ServiceSpec, id: &str) -> Result<(), Error> {
        match self.observe_ollama(spec, id).await {
            Err(Error::PartialRuntime) if id.is_empty() => Ok(()),
            other => other.map(|_| ()),
        }
    }
    pub async fn bound_ollama(&self, id: &str, endpoint: &str) -> Result<Service, Error> {
        let parts: Vec<_> = id.split('/').collect();
        if parts.len() != 3 || parts.iter().any(|s| s.is_empty()) {
            return Err(Error::Conflict("invalid Ollama physical binding"));
        }
        let container = self
            .container(parts[1])
            .await?
            .ok_or(Error::Conflict("bound Ollama container is unobservable"))?;
        let config = container.config.ok_or(ObservationError::Incomplete)?;
        let labels = config.labels.ok_or(ObservationError::Incomplete)?;
        let host = container.host_config.ok_or(ObservationError::Incomplete)?;
        super::Models::new(endpoint)?;
        let bind = endpoint
            .strip_prefix("http://")
            .and_then(|v| v.strip_suffix("/v1"))
            .ok_or(ObservationError::Incomplete)?;
        let spec = ServiceSpec {
            name: container
                .name
                .ok_or(ObservationError::Incomplete)?
                .trim_start_matches('/')
                .into(),
            owner: labels
                .get(OWNER_LABEL)
                .ok_or(ObservationError::Incomplete)?
                .clone(),
            generation: labels
                .get(GENERATION_LABEL)
                .ok_or(ObservationError::Incomplete)?
                .clone(),
            image: config.image.ok_or(ObservationError::Incomplete)?,
            network: host.network_mode.ok_or(ObservationError::Incomplete)?,
            bind_address: bind.into(),
        };
        self.observe_ollama(&spec, id)
            .await?
            .ok_or(Error::Conflict("bound Ollama parent is absent"))
    }
    pub async fn ensure_ollama(&self, spec: &ServiceSpec, id: &str) -> Result<Service, Error> {
        let observed = match self.observe_ollama(spec, id).await {
            Err(Error::PartialRuntime) if id.is_empty() => None,
            other => other?,
        };
        if observed.is_none() {
            if !id.is_empty() {
                return Err(Error::Conflict(
                    "bound Ollama unavailable; recreation forbidden",
                ));
            }
            if self.image(&spec.image).await?.is_none() {
                self.pull_image(&spec.image).await?;
            }
            if self.volume(&spec.volume()).await?.is_none() {
                self.api
                    .create_volume(VolumeCreateRequest {
                        name: Some(spec.volume()),
                        driver: Some("local".into()),
                        labels: Some(spec.labels()?),
                        ..Default::default()
                    })
                    .await
                    .map_err(|error| remote(&error))?;
            }
            let volume = self
                .volume(&spec.volume())
                .await?
                .ok_or(ObservationError::Incomplete)?;
            spec.verify_volume(&volume)?;
            if self.container(&spec.name).await?.is_some() {
                return Err(Error::Conflict(
                    "Ollama appeared during create; reapply to reconcile",
                ));
            }
            self.api
                .create_container(
                    Some(CreateContainerOptions {
                        name: Some(spec.name.clone()),
                        ..Default::default()
                    }),
                    spec.container()?,
                )
                .await
                .map_err(|error| remote(&error))?;
        }
        let observed = self
            .observe_ollama(spec, id)
            .await?
            .ok_or(ObservationError::Incomplete)?;
        if !observed.running {
            let container = observed
                .id
                .split('/')
                .nth(1)
                .ok_or(ObservationError::Incomplete)?;
            self.api
                .start_container(container, None)
                .await
                .map_err(|error| remote(&error))?;
        }
        self.observe_ollama(spec, &observed.id)
            .await?
            .ok_or(ObservationError::Incomplete.into())
    }
}
#[cfg(all(test, unix))]
#[path = "service_tests.rs"]
mod tests;
